### PR TITLE
frontend calls decision service to log interview into mongo

### PR DESCRIPTION
**Pull request by brian-digital**

### CODE REVIEW

File: app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+case class PersonalService(workerSentActualSubstitute: Option[String] = None, workerPayActualSubstitute: Option[String] = None,
+                     possibleSubstituteRejection: Option[String] = None, possibleSubstituteWorkerPay: Option[String] = None,
+                     wouldWorkerPayHelper: Option[String] = None)
+
+
+object PersonalService {
+  def apply(interview: Map[String, Map[String, String]]): PersonalService = {
+    interview.get("personalService").fold[PersonalService]
+      {PersonalService()}
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala]: first part of the fold is for an empty option

File: app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+case class LogInterview(version: String, compressedInterview: String, route: String, decision:String, count: Option[String], setup: Setup,
+                    exit: Exit, personalService: Option[PersonalService], control: Option[Control], financialRisk: Option[FinancialRisk],
+                    partAndParcel: Option[PartAndParcel], completed: DateTime)
+
+case class Setup(endUserRole: String, hasContractStarted: String, provideServices: String)
+
+case class Exit(officeHolder: String)
+
+
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala]: Added default vaules

File: app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+case class PersonalService(workerSentActualSubstitute: Option[String] = None, workerPayActualSubstitute: Option[String] = None,
+                     possibleSubstituteRejection: Option[String] = None, possibleSubstituteWorkerPay: Option[String] = None,
+                     wouldWorkerPayHelper: Option[String] = None)
+
+
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala]: Added an object that can be built from an interview

File: app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    val exit = Exit(
+      interview("exit")("officeHolder")
+    )
+
+    val setup = Setup(
+      interview("setup")("endUserRole"),
+      interview("setup")("hasContractStarted"),
+      interview("setup")("provideServices")
+    )
+
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala]: Refactors to this with logic for PersonalService in its apply method

File: app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import org.joda.time.DateTime
+import play.Logger
+import uk.gov.hmrc.offpayroll.typeDefs.Interview
+import uk.gov.hmrc.offpayroll.util.{ClusterAndQuestion, CompressedInterview}
+
+/**
+  * Created by brianheathcote on 15/07/2017.
+  */
+
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala]: Does not need to be a Package Object

File: app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      interview("setup")("endUserRole"),
+      interview("setup")("hasContractStarted"),
+      interview("setup")("provideServices")
+    )
+
+    def createPersonalService(): PersonalService = {
+      PersonalService(interview)
+    }
+
+
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala]: I would propose to refactor the rest of the methods on this object to follow the pattern for PersonalService

File: app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    def getRoute(route: String): String = {
+      if (route == "soleTrader") "ESI" else "IR35"
+    }
+
+    LogInterview(
+      decisionRequest.version,
+      compressedInterview,
+      getRoute(setup.provideServices),
+      decision.result,
+      Option.empty,
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/LogInterviewBuilder.scala]: None

File: app/uk/gov/hmrc/offpayroll/models/package.scala [hunk 8]
```diff
@@ -16,23 +16,45 @@
...
+  val dateFormat = "yyyy-MM-dd HH:mm:ss"
+
   implicit val decideRequestFormatter: Format[DecisionRequest] = Json.format[DecisionRequest]
   implicit val decideResponseFormatter: Format[DecisionResponse] = Json.format[DecisionResponse]
   implicit val hc = HeaderCarrier()
 
 
+  implicit val dateWrites = Writes.jodaDateWrites(dateFormat)
+  implicit val dateReads = Reads.jodaDateReads(dateFormat)
+
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/package.scala]: Propose to refactor these names to real English

File: app/uk/gov/hmrc/offpayroll/models/package.scala [hunk 9]
```diff
@@ -16,23 +16,45 @@
...
+  implicit val sFormat = Json.format[Setup]
+  implicit val eFormat = Json.format[Exit]
+  implicit val psFormat = Json.format[PersonalService]
+  implicit val cFormat = Json.format[Control]
+  implicit val frFormat = Json.format[FinancialRisk]
+  implicit val ppFormat = Json.format[PartAndParcel]
+  implicit val iFormat = Json.format[LogInterview]
```
Comment (peteanning) [inline app/uk/gov/hmrc/offpayroll/models/package.scala]: Can we include a spec that test the actual value of the Json that an LogInterview can be read and written to

File: test/uk/gov/hmrc/offpayroll/models/LogInterviewBuilderSpec.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+package uk.gov.hmrc.offpayroll.models
+
+import org.scalatest.{FlatSpec, Matchers}
+
+/**
+  * Created by brianheathcote on 15/07/2017.
+  */
+
+class LogInterviewBuilderSpec extends FlatSpec with Matchers  {
```
Comment (peteanning) [inline test/uk/gov/hmrc/offpayroll/models/LogInterviewBuilderSpec.scala]: Needs more tests. Should Check for PersonalService, Control, FinancialRisk etc etc

File: test/uk/gov/hmrc/offpayroll/models/LogInterviewBuilderSpec.scala [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  it should "create and populate a setup object from the interview passed in the decisionRequest" in {
+    logInterview.setup.endUserRole shouldBe "endClient"
+    logInterview.setup.hasContractStarted shouldBe "Yes"
+    logInterview.setup.provideServices shouldBe "partnership"
+  }
+
+  it should "contain a valid PersonalService section" in {
+    val personalService = logInterview.personalService
+    personalService shouldBe defined
+    personalService.get.workerSentActualSubstitute shouldBe defined
```
Comment (peteanning) [inline test/uk/gov/hmrc/offpayroll/models/LogInterviewBuilderSpec.scala]: Perhaps define seperate specs for the sections

File: app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala (hunk 3 above)
Comment (brian-digital) [inline app/uk/gov/hmrc/offpayroll/models/AnalyticsModel.scala]: Optional objects follow pattern for PersonalService

File: test/uk/gov/hmrc/offpayroll/models/LogInterviewBuilderSpec.scala (hunk 10 above)
Comment (brian-digital) [inline test/uk/gov/hmrc/offpayroll/models/LogInterviewBuilderSpec.scala]: I left all the tests in the one spec as it's reasonably understandable and not sure that creating a separate spec for each object would improve it massively

**Pull request merged**